WMCO 10.16.0 release notes (#75198)
openshift/openshift-docs: windows_containers/windows-containers-release-notes-10-16-x.adoc

=== New features and improvements
[id="wmco-10-16-0-node-certificates"]

[id="wmco-10-16-0-disconnected"]
==== WMCO is now supported in restricted networks

The WMCO is now supported in environments with restricted networks, which is a cluster that is intentionally impeded from reaching the internet, also known as disconnected or "air-gapped" clusters.

// For more information, see xref:  windows_containers/enabling-windows-container-workloads.adoc#wmco-disconnedted-cluster_enabling-windows-container-workloads[Using Windows containers in a disconnected cluster].

[id="wmco-10-16-0-mirrored"]
==== WMCO can pull images from mirrored registries

The WMCO can now use both *ImageDigestMirrorSet* (IDMS) and *ImageTagMirrorSet* (ITMS) objects to pull images from mirrored registries.

// For more information, see xref:	../windows_containers/enabling-windows-container-workloads.adoc#images-configuration-registry-mirror_enabling-windows-container-workloads[Understanding image registry repository mirroring]

[id="wmco-10-16-0-filesystem"]
==== *Filesystem* metrics now display for Windows nodes

> The *Filesystem* metrics are now available for Windows nodes in the *Utilization* tile of the *Node details* page in the {product-title} web console. You can xref:../observability/monitoring/managing-metrics.adoc#about-querying-metrics_managing-metrics[query the metrics] by running Prometheus Query Language (PromQL) queries.  The charts previously reported *No datapoints found*.

Reviewer: ```suggestion
The *Filesystem* metrics are now available for Windows nodes in the *Utilization* tile of the *Node details* page in the {product-title} web console. You can xref:../observability/monitoring/managing-metrics.adoc#about-querying-metrics_managing-metrics[query the metrics] by running Prometheus Query Language (PromQL) queries. The charts previously reported *No datapoints found*.
```
Omit extra space

[id="wmco-10-16-0-cpu-memory"]
==== Pod CPU and memory metrics now display for the pods on Windows nodes

The CPU and memory usage metrics are now available for Windows pods on the *Pods* and *Pod details* pages in the {product-title} web console. You can xref:../observability/monitoring/managing-metrics.adoc#about-querying-metrics_managing-metrics[query the metrics] by running PromQL queries. The chart previously reported *No datapoints found*.

[id="wmco-10-16-0-operator-kube"]
==== Kubernetes upgrade

The WMCO now uses Kubernetes 1.29.

[id="wmco-10-16-0-bug-fixes"]
=== Bug fixes

Because the WICD service account was missing a required secret, the WMCO was unable to properly configure Windows nodes in a Nutanix cluster. With this fix, the WMCO creates a long-lived token secret for the WICD service account. As a result, the WMCO is able to configure a Windows node on Nutanix. (link:https://issues.redhat.com/browse/OCPBUGS-22680[*OCPBUGS-22680*])

> Previously, the WMCO performed a sanitization step that incorrectly replaced commas with semicolons in the user's cluster-wide proxy configuration. This caused Windows to ignore the values set in the `noProxy` environment variable. As a consequence, the WMCO incorrectly sent traffic through the proxy for the endpoints specified in the `no-proxy` parameter. With this fix, the sanitization step that replaced commas with semicolons was removed. As a result, web requests from a Windows node to a cluster-internal endpoint or an endpoint that exists in the `no-proxy` parameter does not go through the proxy. (link:https://issues.redhat.com/browse/OCPBUGS-24264[*OCPBUGS-24264*])

Reviewer: ```suggestion
Previously, the WMCO performed a sanitization step that incorrectly replaced commas with semicolons in a user's cluster-wide proxy configuration. This behavior caused Windows to ignore the values set in the `noProxy` environment variable. As a consequence, the WMCO incorrectly sent traffic through the proxy for the endpoints specified in the `no-proxy` parameter. With this fix, the sanitization step that replaced commas with semicolons was removed. As a result, web requests from a Windows node to a cluster-internal endpoint or an endpoint that exists in the `no-proxy` parameter do not go through the proxy. (link:https://issues.redhat.com/browse/OCPBUGS-24264[*OCPBUGS-24264*])
```
- `a user` because you're referring to a generic user and not a specific user
- [Per ISG guidance](https://www.ibm.com/docs/en/ibm-style?topic=grammar-pronouns#ambiguous-pronoun-references), make sure pronoun references are clear by adding a reference for the pronoun "this". "this" what?
- Fix subject verb disagreement `...web requests... do not go`

// Copied from 4.15

> Previously, because of bad logic in the networking configuration script, the WMCO was incorrectly reading carriage returns in the containderd CNI configuration file as changes, and identified the file as modified. This caused the CNI configuration to be unnecessarily reloaded, potentially resulting in container restarts and brief network outages. With this fix, the WMCO now reloads the CNI configuration only when the CNI configuration is actually modified. (link:https://issues.redhat.com/browse/OCPBUGS-2887[*OCPBUGS-2887*])

Reviewer: ```suggestion
Previously, because of bad logic in the networking configuration script, the WMCO was incorrectly reading carriage returns in the containderd CNI configuration file as changes, and identified the file as modified. This behavior caused the CNI configuration to be unnecessarily reloaded, potentially resulting in container restarts and brief network outages. With this fix, the WMCO now reloads the CNI configuration only when the CNI configuration is actually modified. (link:https://issues.redhat.com/browse/OCPBUGS-2887[*OCPBUGS-2887*])
```
- [Per ISG guidance](https://www.ibm.com/docs/en/ibm-style?topic=grammar-pronouns#ambiguous-pronoun-references), make sure pronoun references are clear

// Copied from 4.12
Previously, because of routing issues present in Windows Server 2019, under certain conditions and after more than one hour of running time, workloads on Windows Server 2019 could have experienced packet loss when communicating with other containers in the cluster. This fix enables Direct Server Return (DSR) routing within kube-proxy. As a result, DSR now causes request and response traffic to use a different network path, circumventing the bug within Windows Server 2019. (link:https://issues.redhat.com/browse/OCPBUGS-26761[*OCPBUGS-26761*])

// Copied from 4.15

> Previously, the kubelet on Windows nodes was unable to authenticate with private Elastic Container Registries (ECR) registries. Because of this error, the kubelet was not able to pull images from these registries. With this fix, the kubelet is able to pull images from these registries as expected. (link:https://issues.redhat.com/browse/OCPBUGS-26602[*OCPBUGS-26602*])

Reviewer: ```suggestion
Previously, the kubelet on Windows nodes was unable to authenticate with private Elastic Container Registries (ECR) registries. Because of this error, the kubelet was not able to pull images from these registries. With this fix, the kubelet is able to pull images from these registries as expected. (link:https://issues.redhat.com/browse/OCPBUGS-26602[*OCPBUGS-26602*])
```
- I'm wondering if there's a way to reword to avoid saying `...Registries (ECR) registries`?

// Copied from 4.15

> Previously, on Azure clusters the WMCO would check if an external Cloud Controller Manager (CCM) was being used on the cluster. If a CCM is being used, the Operator would adjust configuration logic accordingly. Because the status condition that the WMCO used to check for the CCM was removed, the WMCO proceeded as if a CCM was not in use. This fix removes the check. As a result, the WMCO always configures the required logic on Azure clusters. (link:https://issues.redhat.com/browse/OCPBUGS-31626[*OCPBUGS-31626*])

Reviewer: ```suggestion
Previously, on Azure clusters the WMCO would check if an external Cloud Controller Manager (CCM) was being used on the cluster. If a CCM was being used, the Operator would adjust configuration logic accordingly. Because the status condition that the WMCO used to check for the CCM was removed, the WMCO proceeded as if a CCM was not in use. This fix removes the check. As a result, the WMCO always configures the required logic on Azure clusters. (link:https://issues.redhat.com/browse/OCPBUGS-31626[*OCPBUGS-31626*])
```
- Make tenses consistent

// Copied from 4.15

> Previously, the WMCO was logging error messages when any command being run through an SSH connection to a Windows instance failed. This was incorrect behavior because some commands are expected to fail. For example, when the WMCO reboots a node, the Operator runs PowerShell commands on the instance until they fail, meaning the SSH connection rebooted as expected. With this fix, only actual errors are now logged. (link:https://issues.redhat.com/browse/OCPBUGS-20255[*OCPBUGS-20255*])

Reviewer: ```suggestion
Previously, the WMCO logged error messages when a command that was run through an SSH connection to a Windows instance failed. This behavior was incorrect because some commands are expected to fail. For example, when the WMCO reboots a node, the Operator runs PowerShell commands on the instance until they fail, meaning the SSH connection rebooted as expected. With this fix, only actual errors are now logged. (link:https://issues.redhat.com/browse/OCPBUGS-20255[*OCPBUGS-20255*])
```
- General tightening up

> Previously, after rotating the `kube-apiserver-to-kubelet-client-ca` certificate, the contents of the `kubetl-ca.crt` file on Windows nodes was not being populated correctly. With this fix, after certificate rotation, the `kubetl-ca.crt` file contains the proper certificates. (link:https://issues.redhat.com/browse/OCPBUGS-22237[*OCPBUGS-22237*])

Reviewer: ```suggestion
Previously, after rotating the `kube-apiserver-to-kubelet-client-ca` certificate, the contents of the `kubetl-ca.crt` file on Windows nodes was not populated correctly. With this fix, after certificate rotation, the `kubetl-ca.crt` file contains the correct certificates. (link:https://issues.redhat.com/browse/OCPBUGS-22237[*OCPBUGS-22237*])
```
- I think `correct` is more precise than `proper` here

Previously, because of a missing DNS suffix in the kubelet host name on instances that are part of a Windows AD domain controller, the cloud provider failed to find VMs by name. With this fix, the DNS suffix is now included in the host name resolution. As a result, the WMCO is able to configure and join Windows instances that are part of AD domain controller. (link:https://issues.redhat.com/browse/OCPBUGS-34758[*OCPBUGS-34758*])

> Previously, registry certificates provided to the cluster by the user were not being loaded into the Windows trust store on each node. As a consequence, image pulls from a mirror registry were failing, because a self-signed CA is required. With this fix, registry certificates are loaded into the Windows trust store on each node. As a result, images can be pulled from mirror registries with self-signed CAs. (link:https://issues.redhat.com/browse/OCPBUGS-36408[*OCPBUGS-36408*])

Reviewer: ```suggestion
Previously, registry certificates provided to the cluster by a user were not loaded into the Windows trust store on each node. As a consequence, image pulls from a mirror registry failed, because a self-signed CA is required. With this fix, registry certificates are loaded into the Windows trust store on each node. As a result, images can be pulled from mirror registries with self-signed CAs. (link:https://issues.redhat.com/browse/OCPBUGS-36408[*OCPBUGS-36408*])
```
- Change to indefinite article `a user` to indicate a generic rather than a specific user
